### PR TITLE
Feat/dynamic fees

### PR DESCRIPTION
**Pull request by quasisamurai**
ICQ relayer must support dynamic gas prices via feemarket so we implemented this behavior:

- Gas prices by default always queried from feemarket
- If query fails for some reason -> relayer uses default gp from config

Also two additional env vars introduced:

- GAS_MULTIPLIER: multiplies queried/default GP by given coeff (basically a tips for validators)
-  MAX_GAS: if queried & previously multiplied gas exceeds this limit, set gas equal to this limit. 

int tests pr
https://github.com/neutron-org/neutron-integration-tests/pull/325

### CODE REVIEW

File: internal/submit/tx_sender.go [hunk 1]
```diff
@@ -114,9 +123,15 @@ func (txs *TxSender) Send(ctx context.Context, msgs []sdk.Msg) (string, error) {
 	txs.lock.Lock()
 	defer txs.lock.Unlock()
 
+	gasPrice, err := txs.getGasPrice(ctx)
```
Comment (pr0n00gler) [inline internal/submit/tx_sender.go:126]: I don't quite understand, you are saying:
> If query fails for some reason -> relayer uses default gp from config

But the logic does the exact opposite - it does return an error that query failed and does not uses default gas prices.

**Review by pr0n00gler — CHANGES_REQUESTED** [review]

File: internal/submit/tx_sender.go (hunk 1 above)
Comment (quasisamurai) [reply internal/submit/tx_sender.go:127]: this logic implemented inside `getGasPrice` func
Снимок экрана 2024-06-18 в 14 18 49

**Comment by quasisamurai** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/9582405898/job/26428850191

File: internal/submit/tx_sender.go [hunk 2]
```diff
@@ -213,6 +229,72 @@ func (txs *TxSender) queryAccount(ctx context.Context, address string) (*authtyp
...
+}
+
+func (txs *TxSender) multiplyGas(gas cosmossdk_io_math.LegacyDec) (string, error) {
+	floatGas, err := gas.Float64()
+	if err != nil {
+		return "", err
+	}
+
+	multipliedGas := txs.gasPriceMultiplier * floatGas
+	if err != nil {
```
Comment (joldie777) [inline internal/submit/tx_sender.go:268]: Why do we need err check here?

File: internal/submit/tx_sender.go [hunk 3]
```diff
@@ -22,29 +24,34 @@ import (
...
 	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	feemarkettypes "github.com/skip-mev/feemarket/x/feemarket/types"
 
 	"github.com/neutron-org/neutron-query-relayer/internal/config"
 )
 
 const (
 	accountQueryPath             = "/cosmos.auth.v1beta1.Query/Account"
 	simulateQueryPath            = "/cosmos.tx.v1beta1.Service/Simulate"
+	getPricesQueryPath           = "/feemarket.feemarket.v1.Quuery/GasPrice"
```
Comment (swelf19) [inline internal/submit/tx_sender.go:35]: looks like an error, does dynamic prices work at all?
```suggestion
	getPricesQueryPath           = "/feemarket.feemarket.v1.Query/GasPrice"
```

File: internal/submit/tx_sender.go [hunk 4]
```diff
@@ -213,6 +229,72 @@ func (txs *TxSender) queryAccount(ctx context.Context, address string) (*authtyp
...
+
+// getGasPrice tries to query dynamic price for denom provided in config.
+// If successful:
+// 1) multiply gas by gasPriceMultiplier (tip validators)
+// 2) if result is bigger than maxGasPrice -> return max gas
+// if query fails for some reason:
+// func returns default gas price[s] from config as well
+func (txs *TxSender) getGasPrice(ctx context.Context) (string, error) {
+	gasPrice, err := txs.queryDynamicPrice(ctx)
+	if err != nil {
```
Comment (swelf19) [inline internal/submit/tx_sender.go:286]: lets log the error here

File: internal/submit/tx_sender.go [hunk 5]
```diff
@@ -213,6 +229,72 @@ func (txs *TxSender) queryAccount(ctx context.Context, address string) (*authtyp
 	return &account, nil
 }
 
+func (txs *TxSender) queryDynamicPrice(ctx context.Context) (cosmossdk_io_math.LegacyDec, error) {
+	request := feemarkettypes.GasPriceRequest{Denom: txs.denom}
+	req, err := request.Marshal()
+	if err != nil {
+		return cosmossdk_io_math.LegacyZeroDec(), fmt.Errorf("error marshalling query gas prices request for denom=%s: %w", txs.denom, err)
+	}
+	simQuery := abci.RequestQuery{
```
Comment (swelf19) [inline internal/submit/tx_sender.go:238]: actually you do not use the structure, we can get rid of it

**Review by swelf19 — COMMENTED** [review]
Let's set default gas price with a really low value in integration tests to  see if dynamic prices really works

**Review by joldie777 — DISMISSED** [review]

**Review by swelf19 — APPROVED** [review]

**Comment by pr0n00gler** [comment]
Please, resolve conflicts

**Review by pr0n00gler — APPROVED** [review]

**Review by joldie777 — APPROVED** [review]

**Comment by quasisamurai** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/9798300917

**Pull request merged**